File: src/main/java/hzt/aoc/day13/Part1ShuttleSearch.java
package hzt.aoc.day13;

import hzt.collections.ListX;
import hzt.sequences.Sequence;

import java.util.List;
import java.util.Map;

public class Part1ShuttleSearch extends Day13Challenge {

    public Part1ShuttleSearch() {
        super("part 1",
                "What is the ID of the earliest bus you can take to the airport " +
                        "multiplied by the number of minutes you'll need to wait for that bus?");
    }

    @Override
    protected String solve(List<String> inputList) {
        final var listX = ListX.of(inputList);
        int earliestTimestamp = Integer.parseInt(listX.first());

        Map<Integer, Integer> busNumbersToWaitingTimes = Sequence.of(inputList.get(1).split(","))
                .filterNot(s -> s.matches("x"))
                .map(Integer::parseInt)
                .associateWith(busNr -> busNr - earliestTimestamp % busNr)
                .toMutableMap();

        int timeToWaitForEarliestBus = Integer.MAX_VALUE;
        int busNumberBelongingToSmallest = 0;
        for (Map.Entry<Integer, Integer> entry : busNumbersToWaitingTimes.entrySet()) {
            if (entry.getValue() < timeToWaitForEarliestBus) {
                timeToWaitForEarliestBus = entry.getValue();
                busNumberBelongingToSmallest = entry.getKey();
            }
        }
        return String.valueOf(timeToWaitForEarliestBus * busNumberBelongingToSmallest);
    }

}

File: src/main/kotlin/hzt/aoc/day13/Part1ShuttleSearch.kt
package hzt.aoc.day13

import java.util.*
import java.util.stream.Collectors

class Part1ShuttleSearch : Day13Challenge(
    "part 1",
    "What is the ID of the earliest bus you can take to the airport " +
            "multiplied by the number of minutes you'll need to wait for that bus"
) {
    override fun solve(inputList: List<String>): String {
        val earliestTimestamp = inputList[0].toInt()
        val integers = Arrays.stream(inputList[1].split(",".toRegex()).toTypedArray())
            .filter { it != "x" }
            .map(String::toInt).collect(Collectors.toList())
        val busNumbersToWaitingTimes = integers.stream()
            .collect(Collectors.toMap(
                    { busNr -> busNr },
                    { busNr -> busNr - earliestTimestamp % busNr })
            )
        var timeToWaitForEarliestBus = Int.MAX_VALUE
        var busNumberBelongingToSmallest = 0
        for ((key, value) in busNumbersToWaitingTimes) {
            if (value < timeToWaitForEarliestBus) {
                timeToWaitForEarliestBus = value
                busNumberBelongingToSmallest = key
            }
        }
        return (timeToWaitForEarliestBus * busNumberBelongingToSmallest).toString()
    }
}
